### DISCUSSION TITLE
I converted a Svelte Todo component to use StateAdapt, and it doesn't actually feel like much of an improvement.

### DISCUSSION THREAD

**Discussion by Evertt** [state-adapt/state-adapt#88] [General]
I'm not 100% sure how to share the code with you, because I wrote it in typescript inside of a big repo full of experiments. The Svelte REPL doesn't support TypeScript, but without TypeScript the code is much less readable, imo. And I've never used Stackblitz, but I imagine that that's not free.

So I'm sorry, I'm just gonna paste it here.

<details>
<summary>This was the `<script>` part of the original `Todo.svelte` component:</summary>

```ts
import { createEventDispatcher } from "svelte"

type Todo = {
	id: string
	description: string
	completed: boolean
}

export let item: Todo

const dispatch = createEventDispatcher()

type SvelteEvent<E extends Event = Event, T extends EventTarget = Element> =
	E & { currentTarget: EventTarget & T }
	
let editing = false

function removeItem() {
	dispatch("removeItem")
}

function startEdit() {
	editing = true
}

// So this function runs on:keydown and either saves the edit or stops the editing
function handleKeyDown(event: SvelteEvent<KeyboardEvent, HTMLInputElement>) {
	if (event.key === "Enter")
		event.currentTarget.blur()
	else if (event.key === "Escape") 
		editing = false
}

// This function run on:blur and either saves the new description or removes the todo
function updateItemOnBlur(event: SvelteEvent<FocusEvent, HTMLInputElement>) {
	if (!editing || !event.target) return
	const { value } = event.currentTarget
	if (value.length) {
		item.description = value
	} else {
		removeItem()
	}
	editing = false
}
```
</details>

That code, honestly, looks very readable to me and easy to understand and therefor I would imagine also easy to maintain.

<details>
<summary>Now this is the code I ended up with when I tried to use StateAdapt and RxJS</summary>

```ts
// First of all, so many more imports...
import { fromEvent, takeUntilDestroyed } from "$lib/rxjs"
import { adapt, Source } from "$lib/state-adapt"
import { joinAdapters } from "@state-adapt/core"
import { booleanAdapter, stringAdapter } from "@state-adapt/core/adapters"
import { filter, mergeWith, map, tap, merge, share } from "rxjs"
import { toSource } from "@state-adapt/rxjs"

type Todo = {
	id: string
	description: string
	completed: boolean
}

type UIState = {
	editing: boolean
}

type State = Todo & UIState

const adapter = joinAdapters<State>()({
	id: stringAdapter,
	description: stringAdapter,
	completed: booleanAdapter,
	editing: booleanAdapter,
})()

export let item: Todo
export let onRemove: (todo: Todo) => void

let editInput: HTMLInputElement | undefined = undefined

// I made my own version of `fromEvent` to make it work
// with Svelte dom element bindings that start out as undefined
const inputKeyDown$ = fromEvent(() => editInput, "keydown")
const inputBlur$ = fromEvent(() => editInput, "blur")

const startEditing$ = new Source("startEditing$")
const toggleCompleted$ = new Source("toggleCompleted$")

const finishEdit$ = merge(inputKeyDown$, inputBlur$).pipe(
	// don't try to finish editing if we're not editing in the first place
	filter(() => editing),

	// Only finish editing when we've hit enter or the input lost focus
	filter(event => !("key" in event) || event.key === "Enter"),

	// map the payload to the value of the edit field of course
	map(e => e.currentTarget?.value.trim() ?? ""),

	// sharing was necessary, otherwise events got dispatched multiple time
	share({ resetOnRefCountZero: true }),
)

const saveEdit$ = finishEdit$.pipe(
	// Save the new description if it's not empty
	filter(description => description.trim().length > 0),
	toSource("saveEdit$"),
)

finishEdit$.pipe(
	// However, if the new description IS empty
	filter(description => description.trim().length === 0),
	map(() => item),
	takeUntilDestroyed(),
// Then we invoke the `onRemove` callback
).subscribe(onRemove)

const stopEditing$ = inputKeyDown$.pipe(
	// We can stop editing if the user hit escape
	filter(event => event.key === "Escape"),

	// Or finished another way
	mergeWith(finishEdit$),
	map(() => ({ id: item.id })),
	share({ resetOnRefCountZero: true }),
	toSource("stopEditing$"),
)

const { editing$ } = adapt({ ...item, editing: false }, {
	adapter,
	sources: ({ state$ }) => ({
		setDescription: saveEdit$,
		setEditingTrue: startEditing$,
		setEditingFalse: stopEditing$,
		toggleCompleted: toggleCompleted$,

		// And then here we simply synchronize
		// this local UI state with the actual todo item
		noop: state$.pipe(
			filter(({ editing }) => !editing),
			tap(state => {
				item.description = state.description
				item.completed = state.completed
			}),
			map(() => ({ type: "saveTodo$", payload: item } as const))
		),
	})
})

// I wanted to just use `$editing$` everywhere instead of `editing`
// but for some reason that made TypeScript hallucinate errors everywhere
let editing = false
$: editing = $editing$
```

I cut out the templating and styling code, because those were nearly identical and functionally these are both identical too.
Now I don't think the version with StateAdapt + RxJS is necessarily bad, but it is much much longer and I'm not yet convinced that that verbosity it bringing any extra clarity with it.

Did I use StateAdapt and RxJS wrong? Could it have been more concise?

**Comment by mfp22** [comment]
Can you share the Svelte code?

**Reply by mfp22** [reply]
Oh nm, I see it now

**Comment by mfp22** [comment]
I'll start with my initial thoughts:

1. `share` is a bad sign. I can't say yet what the alternative is, but I don't think I've ever had to use `share()` on a source. 
2. The `.subscribe(onRemove)` really sucks. Unfortunately Svelte doesn't let you output from observables like Angular. Usually it doesn't matter, but when state is managed in a child component but needs to be reflected in a parent component, this can happen. It's kind of an upside-down relationship. I did this as well when I implemented StateAdapt in TodoMVC. I don't know if I would architect it this way if I started from scratch. Actually I'm pretty sure I wouldn't. If the parent needs to know about an event, it should completely own it. 
3. `noop` should be avoided. All side-effects should be avoided. In this case you are mutating an input, which is not reactive. Again, passing state from child to parent sucks, but it would be preferable to at least separate it out of the store. The intention behind my "sink" idea is to define stuff that isn't defined anywhere else. It's _almost_ declarative. But in this case, it has to pass state to the parent component, so it's a concern from somewhere else cluttering up the store. I would prefer a manual subscribe outside of the store, and have it call a callback/output. 
4. I'm not sure why you combined things into `UI State`. Are all of those states necessarily related? I will look at this more.
5. You added an awful lot of comments to the RxJS version, some of them pointless (like `// map the payload to the value of the edit field of course`, and you didn't put `{}` around the `if` statements in the Svelte version. Were you trying to artificially exacerbate the difference?

You don't have to use StateAdapt for everything, or even anything, if you don't want to. It's extremely important for you to understand the difference between imperative and declarative code, which you might, but the entire point of StateAdapt isn't to control everything but to make reactive programming easier.

> That code, honestly, looks very readable to me and easy to understand and therefor I would imagine also easy to maintain.

This makes me think you chose StateAdapt for a reason other than declarative code. All spaghetti code looks simple when it's small. When it gets big, some developers can't even imagine a different way, so they don't learn anything. Other developers gain an appreciation for declarative code. 

If I could recommend anything, it would be to read the first half of this article: https://dev.to/this-is-angular/progressive-reactivity-in-angular-1d40

After that, count the number of imperative statements in your code that you shared. I got 7 for regular Svelte and 3 with StateAdapt. 

Now I'll start to think about how to start with the Svelte version and see if StateAdapt can improve it, or if it's doomed to spaghetti-ness by the nature of this component.

**Comment by mfp22** [comment]
Okay, let's walk through the Svelte code and make it more reactive. 

```typescript
  import { createEventDispatcher } from "svelte"

  type Todo = {
    id: string
    description: string
    completed: boolean
  }

  export let item: Todo

  const dispatch = createEventDispatcher()

  type SvelteEvent<E extends Event = Event, T extends EventTarget = Element> =
    E & { currentTarget: EventTarget & T }
    
  let editing = false

  // Just use inline like on:click={() => dispatch('removeItem')}
  // function removeItem() {
  //   dispatch("removeItem")
  // }

  // Just set inline like on:click={() => editing = true}
  // function startEdit() {
  //   editing = true
  // }
```

 I think blur is an unfortunate browser API that is incredibly difficult to manage declaratively. See https://twitter.com/mfpears/status/1747284170018312402

So, unfortunately, this should probably remain an event handler:

```typescript
  // So this function runs on:keydown and either saves the edit or stops the editing
  function handleKeyDown(event: SvelteEvent<KeyboardEvent, HTMLInputElement>) {
    if (event.key === "Enter")
      event.currentTarget.blur()
    else if (event.key === "Escape") 
      editing = false
  }
```

Why don't we want to set blur and `editing = false` for both of these? Actually `editing = false` is already downstream from blur. The rule of reactivity is to move everything downstream that can be downstream. It minimizes the risk of inconsistencies.

So actually only 1 thing needs to happen, so this could probably be done inline as well:
```typescript
on:keydown={event => ['Enter', 'Escape'].includes(event.key) && event.currentTarget.blur()}
```

 The reason I'm being liberal with inlining is because in declarative code, event handlers should be minimized at least if they can't be eliminated completely. 

```typescript
  // This function run on:blur and either saves the new description or removes the todo
  function updateItemOnBlur(event: SvelteEvent<FocusEvent, HTMLInputElement>) {
    if (!editing || !event.target) return
    const { value } = event.currentTarget
    if (value.length) {
      item.description = value
    } else {
      removeItem()
    }
    editing = false
  }
```

We have an event, so my first inclination is to make it a Source or Subject. 

Actually you know what? Why not make the output an observable? Don't know exactly how this works in Svelte, but I'll try.
  
```typescript
  let editInput: HTMLInputElement | undefined = undefined
  export const blur$ = fromEvent(() => editInput, 'blur').pipe(toSource('blur$'))

  const editingStore = adapt(false, {
    adapter: booleanAdapter,
    sources: {
      setFalse: blur$,
    }
  })
  const editing$ = editingStore.state$ // Future: support for signals ottb
  $: editing = $editing$
```

Now that editing is a store, the template should call `editingStore.set(true)`.

Now, in the parent component, can react to the same `blur$`. I don't know enough about Svelte to know how exactly this would work or what pitfalls there may be. But it should theoretically be possible. Ideally there would be something like on:exportDefined that you could listen to somehow (create observable), and switch to the actual value (blur$) with a `filter`. But I don't know. 

So, here's the full code:

```typescript
  import {createEventDispatcher} from "svelte"
  import {adapt} from '../adapt.function'
  import {booleanAdapter} from '@state-adapt/core/adapters'

  type Todo = {
    id: string
    description: string
    completed: boolean
  }

  export let item: Todo

  const dispatch = createEventDispatcher()
  
  let editInput: HTMLInputElement | undefined = undefined
  export const blur$ = fromEvent(() => editInput, 'blur').pipe(toSource('blur$'))

  const editingStore = adapt(false, {
    adapter: booleanAdapter,
    sources: {
      setFalse: blur$,
    }
  })
  const editing$ = editingStore.state$
  $: editing = $editing$

  // Move to parent component so it can be reactive to blur$
  function updateItemOnBlur(event: SvelteEvent<FocusEvent, HTMLInputElement>) {
	const { value } = event.currentTarget
	if (value.length) {
		item.description = value
	} else {
		removeItem()
	}
}
```

Does this make sense?

The goal is reactivity by eliminating or minimizing imperative code (including callback functions) and use incrementally complex syntax as needed. State is downstream from events, so rather than passing state as an event, first calculate the state in the parent component and then pass it down to the child.

**Reply by Evertt** [reply]
Okay, I'm gonna take a little time to play around with this until I think I really understand it. One question I already have though is: would you generally advise against using two-way binding? The reason that in my code I used `noop` in the `sources` to sync the store with the `export let item: Todo` was because then I could use two-way binding in the parent component, like this:

```svelte
<ul>
	{#each todos as todo (todo.id)}
		<Todo bind:item={todo} {onRemove} />
	{/each}
</ul>
```

I understand that I could replace this with one-way binding and more event listeners / callback functions for mutating the todo, but I don't see the direct benefit of all that extra syntax quite yet.

### edit

Oh and thank you so much for all the time you've put into writing all this up. I genuinely want to understand how to do declarative reactive programming properly, because even though I only have a very limited understanding of it for now, I do already see that it could have huge potential for leading to much more maintainable code.

**Reply by Evertt** [reply]
> Does this make sense?

Well actually, not completely no. What I think I'm getting from you is that the parent component should really be the one that takes all the responsibility for managing the state of an array of todos. And the `Todo` component should only communicate to the parent using `Source` and / or event dispatches.

But that confuses me, why is it and / or? Why do you say at one point:

> `// Just use inline like on:click={() => dispatch('removeItem')}`

And at another point you recommend to expose a `Source` as a component property:

```ts
export const blur$ = fromEvent(() => editInput, 'blur').pipe(toSource('blur$'))
```

Which suggests that you think the parent component would have something like this:

```svelte
<script lang="ts">
    const todosStore = adapt(initialTodos, {
        adapter: {
            addItem: (todos: ITodo[], item: ITodo) => todos.concat(item),

            removeItem: (todos: ITodo[], item: ITodo) => todos.filter(todo !== item),

            toggleCompleted: (todos: ITodo[], item: ITodo) =>
                todos.map(todo => todo === item ? { ...todo, completed: !todo.completed } : todo),

            setDescription: (todos: ITodo[], { id, description }: ITodo) =>
                todos.map(todo => todo.id === id ? { ...todo, description } : todo),
        },
        sources: {
            setDescription: blur$,
        },
    })

    const todo$ = todosStore.state$

    // I'm not sure how you were imagining this,
    // but I interpreted that you'd think I would
    // bind this blur$ Source to the blur$ that the
    // Todo component exposes as a property
    let blur$ = new Source<ITodo>()
</script>

{#each $todos$ as todo (todo.id)}
    <!-- I hope you see binding `blur$` can't work as there's more than one `blur$` to bind -->
    <Todo item={todo} bind:blur$ on:removeItem={event => todosStore.removeItem(event.detail) />
{/each}
```

And overall it just feels hella inconsistent to `setDescription` using a `Source`, but `removeItem` using an event.
I'm pretty sure I'm just confused about what you really meant. Would you be willing explain this part a little more still?

**Reply by Evertt** [reply]
> 1. `share` is a bad sign. I can't say yet what the alternative is, but I don't think I've ever had to use `share()` on a source.

Well okay, I have noticed that when I check the logs in the redux devtools, then yes I only see all my Actions show up once, exactly how often I would like them to show up. However, when I do my own logging through `tap()` I do see that in certain cases a `Source` runs multiple times. And that's actually very easy to explain. Some sources are used for multiple store reactions downstream.

For example the `enter$` key source is linked to both `editingStore.setFalse()` and indirectly also to `todosStore.setDescription`.
So I guess I don't strictly *need* to `share()` the `Source`, since it's not *that* computationally intensive. But I still just don't like the thought of code running more times than it needs to.

Anyway, I did a new attempt. I'm afraid it's still as long as before, but maybe brevity is not what I should be measuring here.

<details>
<summary>Here is `Todo.svelte`</summary>

```svelte
<script context="module" lang="ts">
	// As you can see, now most of my code lives in the "module" context
	// because then I can import all these sources in the parent component

	import { merge, share } from "rxjs"
	import { adapt, Source, toSource } from "$lib/state-adapt"

	export type Todo = {
		id: string
		description: string
		completed: boolean
	}

	type SvelteFocusEvent = SvelteEvent<FocusEvent, HTMLInputElement>
	type SvelteKeyboardEvent = SvelteEvent<KeyboardEvent, HTMLInputElement>

	const blur$ = new Source<[Todo, SvelteFocusEvent]>("blur$", {
		curried: true,
		ignoreEvents: false,
	})

	const newBlur$ = blur$.pipe(
		map(({ payload: [todo, event] }) => ({
			...todo,
			// I'm just taking the description from the input field
			// otherwise the description just refers to the old todo's description
			description: event.currentTarget?.value ?? ""
		} as Todo)),
		toSource("newBlur$"),
	)

	const keydown$ = new Source<[Todo, SvelteKeyboardEvent]>("keydown$", {
		curried: true,
		ignoreEvents: false,
	})

	const filterKeySource = <K extends string>(keyEvent: Source<[Todo, SvelteKeyboardEvent]>, key: K) =>
		keyEvent.pipe(
			filter(({ payload: [, event] }) => event.key === key),
			map(({ payload: [todo, event] }) => ({
				...todo,
				// Same here, just making sure I take the new description
				description: event.currentTarget?.value ?? ""
			} as Todo)),
			toSource(`${key}$`),
			share({ resetOnRefCountZero: true }),
		)

	const enter$ = filterKeySource(keydown$, "Enter")
	const escape$ = filterKeySource(keydown$, "Escape")

	const finishEditing$ = merge(enter$, newBlur$).pipe(
		toSource("finishEditing$"),
		share({ resetOnRefCountZero: true }),
	)

	export const setDescription$ = finishEditing$.pipe(
		filter(v => v.payload.description.length > 0),
		toSource("setDescription$"),
	)

	const deleteBtnClick$ = new Source<Todo>("deleteBtnClick$", { curried: true })

	const descriptionCleared$ = finishEditing$.pipe(
		filter(v => v.payload.description.length === 0),
		toSource("descriptionCleared$"),
	)

	export const removeItem$ = merge(deleteBtnClick$, descriptionCleared$)
	export const toggleCompleted$ = new Source<Todo>("toggleCompleted$", { curried: true })

	export type SvelteEvent<E extends Event, T extends EventTarget> =
		E & { currentTarget: EventTarget & T }
</script>

<script lang="ts">
	import { tick } from "svelte"
	import { filter, map } from "rxjs"
	import { booleanAdapter } from "@state-adapt/core/adapters"

	export let item: Todo
	
	const dblClick$ = new Source("dblClick$")
	
	const editingStore = adapt(false, {
		adapter: booleanAdapter,
		sources: {
			setFalse: [blur$, escape$, enter$],
			setTrue: dblClick$,
		}
	})

	const editing$ = editingStore.state$
	
	function focusInput(element: HTMLElement) {
		tick().then(() => element.focus())
	}
</script>
	
<li class:completed={item.completed} class:$editing$ class="my-2">
	<div class="view">
		<input class="toggle" type="checkbox" on:change={toggleCompleted$(item)} checked={item.completed} />
		<!-- svelte-ignore a11y-label-has-associated-control -->
		<label on:dblclick={dblClick$}>{item.description}</label>
		<button on:click={deleteBtnClick$(item)} class="px-2 py-1 text-white bg-red-600 rounded">Delete</button>
	</div>
	
	{#if $editing$}
		<div class="input-container">
			<input
				class="edit"
				value={item.description}
				on:keydown={keydown$(item)}
				on:blur={blur$(item)}
				use:focusInput
			/>
		</div>
	{/if}
</li>

<style>
	.completed {
		text-decoration: line-through;
	}
	
	.\$editing\$ {
		border-bottom: 1px solid #ccc;
	}
</style>
```
</details>

<details>
<summary>And this is how the parent component `TodoList.svelte` looks:</summary>

```svelte
<script lang="ts">
	import { adapt } from "$lib/state-adapt"
	import Todo, { setDescription$, removeItem$, toggleCompleted$, type Todo as ITodo } from "./Todo.svelte"

	const initialTodos: ITodo[] = [
		{ id: "1", description: "Learn Svelte", completed: false },
		{ id: "2", description: "Learn SvelteKit", completed: false },
		{ id: "3", description: "Build something awesome", completed: false },
	]
	
	const todosStore = adapt(initialTodos, {
		adapter: {
			addItem: (todos: ITodo[], item: ITodo) =>
				todos.concat(item),

			removeItem: (todos: ITodo[], item: ITodo) =>
				todos.filter(todo => todo.id !== item.id),
			
			toggleCompleted: (todos: ITodo[], item: ITodo) =>
				todos.map(todo => todo === item ? { ...todo, completed: !todo.completed } : todo),

			setDescription: (todos: ITodo[], { id, description }: ITodo) =>
				todos.map(todo => todo.id === id ? { ...todo, description } : todo),
		},
		sources: {
			removeItem: removeItem$,
			toggleCompleted: toggleCompleted$,
			setDescription: setDescription$,
		}
	})

	const todos = todosStore.state$
</script>

<h1>Todos</h1>

<ul>
	{#each $todos as todo (todo.id)}
		<Todo item={todo} />
	{/each}
</ul>
```
</details>

**Reply by mfp22** [reply]
> would you generally advise against using two-way binding?

I included my entire reasoning in my comments. I don't have a general rule against 2-way binding. But if 1 event needs to change 2 top-level states, 2-way binding is impossible without additional imperative code. I would start with 2-way binding and if a 2nd state needs to update, switch to observable + 2 states that reference that observable. 

> `let blur$ = new Source<ITodo>()`

I didn't expect this to be defined in 2 places. I don't know Svelte. I said:

> Now, in the parent component, can react to the same blur$. I don't know enough about Svelte to know how exactly this would work or what pitfalls there may be. But it should theoretically be possible. Ideally there would be something like on:exportDefined that you could listen to somehow (create observable), and switch to the actual value (blur$) with a filter. But I don't know.

You could define `const blur$ = new Source<ITodo>()` and pass it down and just use that in the child component and `.next()` onto it with a manual `on:blur` listener in the template. 

I really don't care about these details. The goal is to make each event only make 1 change upstream, and then have everything be derived from that. The `fromEvent` syntax makes it so not even 1 change is required. That's cool but sort of extra. 

> What I think I'm getting from you is that the parent component should really be the one that takes all the responsibility for managing the state of an array of todos.

I don't have an opinion, but it's good to avoid duplicating state. If you're passing state up to a parent component after a change, it's weird. This is the main point:

> The goal is reactivity by eliminating or minimizing imperative code (including callback functions) and use incrementally complex syntax as needed. State is downstream from events, so rather than passing state as an event, first calculate the state in the parent component and then pass it down to the child.

> `// As you can see, now most of my code lives in the "module" context`

I didn't know about this, but wouldn't these be singletons? So every child instance would be pushing to the same streams? I guess that's why you're passing the payload like `[Todo, SvelteFocusEvent]`. This is the type of thing Angular devs would typically move to a service I guess. 

> However, when I do my own logging through tap() I do see that in certain cases a Source runs multiple times.

Here's the funny thing about sources. You might be the first person to notice this. If you have a source, no matter how many stores you pass it to, it only gets one subscription. I have a central object where I keep track of all the global store state updates that need to happen for a single source. When a store is initialized (has subscribers), for each source i checks if it's already registered, and if so, it just adds to the array of updates to perform without subscribing to it again. I don't use the action type for this, but the source reference itself. So when you use an operator on a source, it returns a new reference, so it gets registered separately and subscribed to independently and set up with its own action in Redux Devtools. So `source$.pipe(delay(1000))` will be guaranteed to be its own unique event. 

It's a little bit weird, but it's why I don't think I've ever used `share` on a source. I wanted the simple cases to be really simple. One common pitfall is in Angular making redundant requests because of subscribing multiple times to the same HTTP observable. In StateAdapt that issue is smoothed over for beginners. Honestly though they should have implemented it with `share` already though. 

StateAdapt's job is basically to solve problems that really should have been solved by either frameworks or RxJS.
